### DISCUSSION TITLE
Perfect Replacement Policy

### DISCUSSION THREAD

**Discussion by nprzybylski** [ChampSim/ChampSim#458] [Q&A]
Hi, I'm a developer and have been trying to implement an optimal replacement policy in ChampSim. The goal is to evict the cache line that will be reused furthest in the future. My approach to this problem involves a first pass analysis of the trace followed by a run of the simulation using the custom policy. I was using the instruction id as an indicator of when each access occurred, but this proved to be unreliable since instructions are processed out of order. I tried reordering the stream of the requests to get a monotonically increasing counter, but wasn't able to successfully segment the accesses. I have a few questions that might help me work through this:

1. Is there any way to use an in-order CPU model in ChampSim?
2. If not, would it make sense to modify the O3_CPU::schedule_instruction() or O3_CPU::do_scheduling() functions to disable the reordering?
3. Is something like this even feasible in ChampSim?
4. Any other thoughts/something I might be missing?

I'm relatively new to ChampSim so I'd appreciate any advice on this topic. Thanks.

**Comment by ngober** [comment]
Over the years, a few different people have suggested that ChampSim have an in-order CPU model, but none of those suggestions turned into pull requests. 

If you would like to embark on this, I think it could certainly be feasible. The `O3_CPU` model separates operation and operand selection: there is a `O3_CPU::fetch_instruction()` that selects a block to fetch and a `O3_CPU::do_fetch_instruction()` that performs the operation. With this, it could be somewhat straightforward to simply apply the `O3_CPU::do_*()` functions to an in-order set of operands.

**Reply by nprzybylski** [reply]
Surely this wouldn't be as simple as changing line 240 in ooo_cpu.cc from

`auto l1i_req_begin = std::find_if(std::begin(IFETCH_BUFFER), std::end(IFETCH_BUFFER), fetch_ready);`
to 
`auto l1i_req_begin = std::begin(IFETCH_BUFFER);`

From what I could gather from the code, each instruction is assigned a status of either `INFLIGHT` or `COMPLETED` to determine whether its execution has finished. If an instruction is marked as completed and has not been fetched yet, then it is ok to fetch. So by removing the `find_if` that checks for these conditions and only selecting the beginning of `IFETCH_BUFFER`, this would fetch instructions as they were read in from the trace? This logically makes sense to me, but there may be something under the hood that I'm missing. Just wanted to make sure.

**Reply by ngober** [reply]
It is not that simple. You have suggested a scheme that is fully in-order, but that has no buffering, and cannot run ahead. You've also restricted it to just in-order fetch, but fetch in ChampSim is already in order. You would need to search though all queues in the core model in program order, in your search skipping candidates only if they are completed (or in progress).

**Reply by nprzybylski** [reply]
I see. Is there a third state other than completed or in progress, or is the default just a null state?

**Reply by nprzybylski** [reply]
Never mind, I see that default is 0, inflight is 1, and completed is 2. Thank you for helping with this.
